### PR TITLE
Fix haskellPackages.fltkhs

### PR DESCRIPTION
**Pull request by chris-martin**
###### Motivation for this change

`haskellPackages.fltkhs` was marked as broken. It apparently requires FLTK >= 1.4 (see https://github.com/deech/fltkhs/issues/93), which we didn't have yet. As far as I can tell, FLTK 1.4 isn't actually released yet, so I didn't think it was appropriate to *replace* the FLTK 1.3 package that we already have, so I added a new version. This fixes the Haskell package.

While I was at it, I also added the dependencies required to enable OpenGL support in fltkhs.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

/cc @deech

### CODE REVIEW

**Review by peti — CHANGES_REQUESTED** [review]
I think that most of those changes should be made in `cabal2nix` to generate a proper Nix expression from the outset rather than to fix it up after the fact.

File: pkgs/development/haskell-modules/configuration-common.nix [hunk 1]
```diff
@@ -694,9 +694,12 @@ self: super: {
   # We cannot build this package w/o the C library from <http://www.phash.org/>.
   phash = markBroken super.phash;
 
-  # https://github.com/deech/fltkhs/issues/16
-  # linking fails because the build doesn't pull in the libGLU_combined libraries
-  fltkhs = markBroken super.fltkhs;
+  fltkhs = let inherit (pkgs.lib) flip foldr; in foldr (f: x: f x) super.fltkhs [
+    (flip appendConfigureFlag "-fopengl")
```
[the remark this answers is not in the document]
Comment (chris-martin) [reply pkgs/development/haskell-modules/configuration-common.nix]: Done - https://github.com/NixOS/cabal2nix/pull/388

**Comment by chris-martin** [comment]
The cabal2nix PR should take care of `appendConfigureFlag "-fopengl"` and `addExtraLibrary super.OpenGLRaw` (which I'm hoping now ends up implied by the flag.

What about `addPkgconfigDepend pkgs.libGLU_combined` and `addSetupDepend pkgs.fltk14`, do those go somewhere in cabal2nix as well?

**Comment by peti** [comment]
> The cabal2nix PR should take care of `appendConfigureFlag "-fopengl"` and `addExtraLibrary super.OpenGLRaw` (which I'm hoping now ends up implied by the flag.

OK. I've re-generated the hackage package set in `haskell-updates` branch @ 8272d0638f9d157ff721a59c573be76d3094adec with the new version.

> 
> What about `addPkgconfigDepend pkgs.libGLU_combined` and `addSetupDepend pkgs.fltk14`, do those go somewhere in cabal2nix as well?

There are two places that might be relevant: the post-process hooks at https://github.com/NixOS/cabal2nix/blob/master/src/Distribution/Nixpkgs/Haskell/FromCabal/PostProcess.hs#L72 allow you to modify the generated Nix expression as you see fit, i.e. to fix up insufficient Cabal files.

Then there is the "name lookup" code in https://github.com/NixOS/cabal2nix/blob/master/src/Distribution/Nixpkgs/Haskell/FromCabal/Name.hs which you can extend to make sure cabal2nix maps system library names used in the Cabal file to Nixpkgs names properly. I'm not sure which of the two is more appropriate your this particular use case. Using post-process hooks to add those dependencies is almost certainly going to work fine, though.

**Review by peti — CHANGES_REQUESTED** [review]
It's probably a good idea to target `haskell-updates` with this PR because it has the newly generated code for `hackage-packages.nix`.

**Comment by peti** [comment]
I cherry-picked the addition of fltkhs 1.4.x in 0aff1d86800570799ac16912fc3cda9e177ebdb6. The necessary fixes in cabal2nix are in https://github.com/NixOS/cabal2nix/commit/210398918c9c82ea56db5c378fb7c5a318e36b59. The updated package set is in 59a8047dcb9916f342cc9c1146ef333d77316bd5. `fltkhs` should now compile fine in the `haskell-updates` branch. I'll merge to `master` soon!

**Comment by chris-martin** [comment]
Awesome, thanks for carrying this through to completion.

**Pull request closed without merging**